### PR TITLE
Define routes

### PR DESCRIPTION
**Pull request by aleksandar-cakic**
Defined the following routes: 
- getProfile: GET request which renders user profile based on userId
- putEditProfile: PUT request which updates the user profile based on userId (user must be authenticated)
- getFriendsList: GET request which returns an array of objects containing information for each friend of current user
- postFriendFollow: POST request to follow a new friend (might remove and keep only PUT later)
- putFriendFollow: PUT request to follow and unfollow a friend 
- getSearchFriends: GET request to search for friends using the unique email 

Started working on queries and controllers. I need to check with backend team on parameters and userId which is why I left comments in code. 

Implemented testing for routes.js, once I finish implementing controllers and queries, I will add tests for those as well

### CODE REVIEW

File: api/routes.js [hunk 1]
```diff
@@ -1,9 +1,28 @@
...
 const controllers = require('./controllers');
-const auth = require('./middleware').auth;
 
-// EXAMPLES (Each route has a controller)
-router.get('/example', auth, controllers.getExample); // Use the auth middleware to authorize requests. Requests must have a valid JWT token from the auth service
-router.post('/example', controllers.setExample);
+/// USER ROUTES ///
+
+// GET request to render user profile information based on provided user ID
+router.get('/user/:userId', controllers.getProfile);
```
Comment (callumreid) [inline api/routes.js]: Will definitely need to work with you in connecting the front and back end of user at some point. These routes seem to cover my operations, which are basically 'get all user info' and 'edit user info'

File: api/routes.js [hunk 2]
```diff
@@ -1,9 +1,28 @@
...
+// PUT request to update the user profile (provided userID must match currently logged in user)
+router.put('/user/:userId/edit', controllers.putEditProfile);
+
+/// FRIEND ROUTES ///
+
+// GET request for rendering a list of all friends for the current user
+// -> returns an array of objects containing information for each friend of current user
+router.get('/user/:userId/friendsList', controllers.getFriendsList);
+
+// POST request to follow a new friend
```
Comment (callumreid) [inline api/routes.js:18]: Curious about having a POST for following but a PUT for unfollowing. I'm sure it makes sense it just struck me that unfollowing was not a delete operation

File: api/controllers.js [hunk 3]
```diff
@@ -2,13 +2,78 @@ const db = require('../service');
...
+
+  getProfile: (req, res) => {
+    if (!req.params.userId) {
+      res.status(400).send('Missing param "userId"');
+      return;
+    }
+    return db.getProfile(req.params.userId)
+      .then((response) => {
+        if (!response || response.rowCount === 0) {
+          res.status(400).send(`Unable to get user profile based on userId: ${req.params.userId}`);
```
Comment (callumreid) [inline api/controllers.js]: Helpful error handling messages 👌👌👌

File: service/index.js [hunk 4]
```diff
@@ -1,4 +1,93 @@
-// Set up you DB here
-const db = {};
+require('dotenv').config();
 
-module.exports = db;
\ No newline at end of file
+const { Pool, Client } = require('pg');
+
+const pool = new Pool({
```
Comment (callumreid) [inline service/index.js:5]: Have not used pool connections before, but I see them on SO often. I'll have to look into their use

File: service/index.js [hunk 5]
```diff
@@ -1,4 +1,93 @@
...
+  database: process.env.POSTGRES_DB,
+  password: process.env.POSTGRES_PASSWORD,
+  port: process.env.POSTGRES_PORT
+});
+
+const getProfile = (userId) => {
+  return pool
+    .connect()
+    .then((client) => {
+      const query = 'SELECT * FROM users WHERE first_name = $1'; // to update userId param
```
Comment (callumreid) [inline service/index.js:17]: Is $1 Postgres specific wildcard syntax or is this hardcoded?

File: service/index.js [hunk 6]
```diff
@@ -1,4 +1,93 @@
...
+      return client.query(query, values);
+    })
+    .catch((err) => {
+      client.release();
+      console.log(err.stack);
+      return null;
+    });
+};
+
+const putEditProfile = (userId, first_name, last_name, age, snack) => {
```
Comment (callumreid) [inline service/index.js:28]: Might need an edit option for animal/species-- if this is problematic could find a workaround in front end

File: tests/api/routes.test.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
+const app = require('../../server.js');
+const supertest = require('supertest');
+const request = supertest(app);
+
+describe('User Endpoints', () => {
```
Comment (callumreid) [inline tests/api/routes.test.js:5]: Ahh way to have tests! Love to see it, these seem really well set up

File: tests/api/routes.test.js [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+    it('should return a 400 status code for invalid userId', async () => {
+      const response = await request.get('/user/Rick');
+
+      expect(response.status).toBe(400);
+      expect(Array.isArray(response.body)).toBe(false);
+      expect(response.text).toBe('Unable to get user profile based on userId: Rick');
+    });
+
+    it('should return a 404 status code for missing userId', async () => {
```
Comment (callumreid) [inline tests/api/routes.test.js]: Way to check not happy path as well

File: tests/api/routes.test.js [hunk 9]
```diff
@@ -67,32 +67,32 @@ describe('Friends Endpoints', () => {
   describe('GET Search Friend', () => {
 
     it('should return a user profile based on email', async () => {
-      const response = await request.get('/searchFriend/joe@mama.com');
+      const response = await request.get('/searchFriend?email=joe@mama.com');
```
Comment (callumreid) [inline tests/api/routes.test.js:70]: 😂

File: api/controllers.js [hunk 10]
```diff
@@ -13,7 +13,7 @@ module.exports = {
     return db.getProfile(req.query.userId)
       .then((response) => {
         if (!response || response.rowCount === 0) {
-          res.status(400).send(`Unable to get user profile based on userId: ${req.query.userId}`);
+          res.status(400).send('Unable to get user profile based on userId');
```
Comment (callumreid) [inline api/controllers.js:16]: surprises me that this was a security issue

**Review by callumreid — APPROVED** [review]
Looks great! Super clean code that is easy to follow. One question as to whether or not we are storing animal/species for a user. I could hack around it for front end, we could just not include it, or we could add it as something to be saved. I should have realized earlier the discrepancy but I did not. Merged.

File: api/routes.js (hunk 1 above)
Comment (aleksandar-cakic) [reply api/routes.js]: I'm free whenever you have time!

File: api/routes.js (hunk 2 above)
Comment (aleksandar-cakic) [reply api/routes.js:18]: I had it as DELETE at first, however, if you are looking to follow, unfollow, follow, unfollow the same person over and over again, PUT makes more sense for updating the two profiles. I'm also thinking of removing POST for follow and just use PUT for follow/unfollow, but I'll have to look into it more

File: service/index.js (hunk 5 above)
Comment (aleksandar-cakic) [reply service/index.js:17]: I believe it's Postgres specific. It's the first item in the values array. At the moment I only have const values = [userId], however, if you had more items, you would have $1, $2, etc for marking them. Think of it as VALUES (?, ?, ?) for mySQL queries

File: service/index.js (hunk 6 above)
Comment (aleksandar-cakic) [reply service/index.js:28]: Should be an easy add, I'll make note of it :)

File: tests/api/routes.test.js (hunk 7 above)
Comment (aleksandar-cakic) [reply tests/api/routes.test.js:5]: Thank you! It took a while to set them up properly :S

File: tests/api/routes.test.js (hunk 9 above)
Comment (aleksandar-cakic) [reply tests/api/routes.test.js:70]: What an animal, this Joe dude

File: api/controllers.js (hunk 10 above)
Comment (aleksandar-cakic) [reply api/controllers.js:16]: Yeah, SonarCloud gave me lots of links explaining how it is a huge XSS vulnerability + ways to refactor. It pointed me to sanitizeHTML npm module, but I think this refactor was faster/easier

**Pull request merged**